src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -1,164 +1,581 @@
+                private readonly ArrayBuilder<LocalSymbol> _temps = ArrayBuilder<LocalSymbol>.GetInstance();
+
+                public DagTempAllocator(SyntheticBoundNodeFactory factory)
+                {
+                    this._factory = factory;
+                }
+
+                public void Dispose()
+                {
+                    _temps.Free();
+                    _map.Free();
+                }
+
+                public BoundExpression GetTemp(BoundDagTemp dagTemp)
+                {
+                    if (!_map.TryGetValue(dagTemp, out var result))
                     {
-                        var constantPattern = (BoundConstantPattern)pattern;
-                        return constantPattern.Update(VisitExpression(constantPattern.Value), constantPattern.ConstantValue);
+                        // PROTOTYPE(patterns2): Not sure what temp kind should be used for `is pattern`.
+                        var temp = _factory.SynthesizedLocal(dagTemp.Type, syntax: _factory.Syntax, kind: SynthesizedLocalKind.SwitchCasePatternMatching);
+                        _map.Add(dagTemp, _factory.Local(temp));
+                        _temps.Add(temp);
+                        result = _factory.Local(temp);
                     }
-                default:
-                    return pattern;
-            }
-        }
 
-        private BoundExpression MakeIsConstantPattern(BoundConstantPattern loweredPattern, BoundExpression loweredInput)
-        {
-            return CompareWithConstant(loweredInput, loweredPattern.Value);
-        }
+                    return result;
+                }
 
-        private BoundExpression MakeIsDeclarationPattern(BoundDeclarationPattern loweredPattern, BoundExpression loweredInput)
-        {
-            Debug.Assert(((object)loweredPattern.Variable == null && loweredPattern.VariableAccess.Kind == BoundKind.DiscardExpression) ||
-                         loweredPattern.Variable.GetTypeOrReturnType() == loweredPattern.DeclaredType.Type);
+                public ImmutableArray<LocalSymbol> AllTemps()
+                {
+                    return _temps.ToImmutableArray();
+                }
+
+                public void AssignTemp(BoundDagTemp dagTemp, BoundExpression value)
+                {
+                    _map.Add(dagTemp, value);
+                }
+            }
 
-            if (loweredPattern.IsVar)
+            private bool LowerDecision(BoundDagDecision decision)
             {
-                var result = _factory.Literal(true);
+                var oldSyntax = _factory.Syntax;
+                _factory.Syntax = decision.Syntax;
+                var result = LowerDecisionCore(decision);
+                _factory.Syntax = oldSyntax;
+                return result;
+            }
 
-                if (loweredPattern.VariableAccess.Kind == BoundKind.DiscardExpression)
+            private bool LowerDecisionCore(BoundDagDecision decision)
```

Comment:
>LowerDecisionCore [](start = 25, length = 17)

Please add a comment about the purpose of this function and how the return should be interpreted.

Reply:
Appears to have been addressed by or in [https://github.com/dotnet/roslyn/pull/23357](https://github.com/dotnet/roslyn/pull/23357).

---
In reply to: [156541223](https://github.com/dotnet/roslyn/pull/23209#discussion_r156541223) [](ancestors = 156541223)